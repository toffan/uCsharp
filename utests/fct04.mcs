int* malloc(int size) {
    int* addr;
    asm {
        LOAD %size
        SUBR Malloc
        STORE %addr
    }
    return addr;
}

int max(int a, int b) {
    int c = 25;
    int* p = malloc(1);
    *p = 3;
    if (a > b) {
        *p = 2;
        return a;
    } else {
        *p = 5;
        return b;
    }
}

int main() {
    int d = 3;
    int c = max(9, 3);
    return d;
}
