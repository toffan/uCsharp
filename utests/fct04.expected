JUMP __malloc
_malloc:
    PUSH 1
    LOAD (1) -1[LB]
    SUBR Malloc
    STORE (1) 3[LB]
    LOAD (1) 3[LB]
    RETURN (1) 1
    RETURN (0) 1
__malloc:

JUMP __max
_max:
    LOADL 25
    LOADL 1
    CALL (SB) _malloc
    LOADL 3
    LOAD (1) -1[ST]
    LOAD (1) 4[LB]
    STOREI (1)
    POP (0) 1
    LOAD (1) -2[LB]
    LOAD (1) -1[LB]
    SUBR IGtr
    SUBR I2B
    JUMPIF (0) ELSE_0
    LOADL 2
    LOAD (1) -1[ST]
    LOAD (1) 4[LB]
    STOREI (1)
    POP (0) 1
    LOAD (1) -2[LB]
    RETURN (1) 2
    POP (0) 0
    JUMP ENDIF_0
ELSE_0:
    LOADL 5
    LOAD (1) -1[ST]
    LOAD (1) 4[LB]
    STOREI (1)
    POP (0) 1
    LOAD (1) -1[LB]
    RETURN (1) 2
    POP (0) 0
ENDIF_0:
    RETURN (0) 2
__max:

JUMP __main
_main:
    LOADL 3
    LOADL 9
    LOADL 3
    CALL (SB) _max

    LOAD (1) 3[LB]
    RETURN (1) 0
    RETURN (0) 0
__main:
    CALL (SB) _main
    HALT
